File: Java/프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());

        for (int[] road : roads) {
            graph.get(road[0]).add(road[1]);
            graph.get(road[1]).add(road[0]);
        }

        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);
        dist[destination] = 0;

        // destination에서 딱 한 번만 BFS
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(destination);
        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int next : graph.get(cur)) {
                if (dist[next] == -1) {          // 방문 체크 겸 거리 갱신
                    dist[next] = dist[cur] + 1;
                    q.offer(next);
                }
            }
        }

        int[] answer = new int[sources.length];
        for (int i = 0; i < sources.length; i++) {
            answer[i] = dist[sources[i]];         // O(1) 조회
        }
        return answer;
    }
}

File: C++/프로그래머스/3/132266. 부대복귀/부대복귀.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

/*
    모두 1로 동일하니, 다익스트라 이런 어려운 로직보다 BFS를 쓸 수 있다.
    
    노드를 사용하는 BFS를 사용하면 빠르게 풀 수 있을 것 같음. 다익스트라를 사용핻 되고.
    단, 길이 50만개라서 2중반복을 사용하는 BFS는 사용 못할 수도 있음
    일단 지역이 1베이스라서 n + 1로 생성해주어야 함.
*/


int bfs(int start, int end, const vector<vector<int>>& map, vector<bool> visited) {
    queue<pair<int, int>> q; // node idx, cost
    q.push({start, 0});
    visited[start] = true;
    
    while (!q.empty()) {
        pair<int, int> cur = q.front();
        q.pop();
        
        if (cur.first == end) return cur.second;
        
        for (int next : map[cur.first]) {
            if (!visited[next]) {
                visited[next] = true;
                q.push({next, cur.second + 1});
            }
        }
    }
    return -1;
}

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination) {
    vector<int> ans;
    vector<vector<int>> map(n + 1);
    vector<bool> visited(n + 1);
    
    for (int i = 0; i < roads.size(); i++) {
        map[roads[i][0]].push_back(roads[i][1]);
        map[roads[i][1]].push_back(roads[i][0]);
    }
    
    for (int i = 0; i < sources.size(); i++) {
        ans.push_back(bfs(sources[i], destination, map, visited));
    }
    
    return ans;
}